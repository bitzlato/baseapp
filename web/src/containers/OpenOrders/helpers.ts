export const getTriggerSign = (ordType: string, ordSide: string) => {
  if (ordType.includes('stop')) {
    if (ordSide === 'buy') {
      return '≥';
    }
    return '≤';
  }

  if (ordType.includes('take')) {
    if (ordSide === 'buy') {
      return '≤';
    }
    return '≥';
  }
};
