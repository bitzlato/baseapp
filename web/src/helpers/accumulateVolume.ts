export const accumulateVolume = (array: any) => {
  const total: number[] = [];
  array
    .map((item: any) => {
      return item[1];
    })
    .reduce((accumulator: any, currentValue: any, currentIndex: any) => {
      total[currentIndex] = Number(accumulator) + Number(currentValue);

      return Number(accumulator) + Number(currentValue);
    }, 0);

  return total;
};
