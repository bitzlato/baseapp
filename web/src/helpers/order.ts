import { OrderType } from 'src/components/Order';

const ORDER_TYPES_WITH_TRIGGER: OrderType[] = [
  'Stop-loss',
  'Take-profit',
  'Stop-limit',
  'Take-limit',
];

export function isTrigger(orderType: OrderType): boolean {
  return ORDER_TYPES_WITH_TRIGGER.includes(orderType);
}

const ORDER_TYPES_WITH_LIMIT: OrderType[] = ['Limit', 'Stop-limit', 'Take-limit'];

export function isLimit(orderType: OrderType): boolean {
  return ORDER_TYPES_WITH_LIMIT.includes(orderType);
}

const TRIGGER_BUY_PRICE_ADJUSTED_TYPES: OrderType[] = ['Stop-loss', 'Take-profit'];

export function isTriggerByPrice(orderType: OrderType): boolean {
  return TRIGGER_BUY_PRICE_ADJUSTED_TYPES.includes(orderType);
}

export function isMarket(orderType: OrderType): boolean {
  return orderType === 'Market';
}
