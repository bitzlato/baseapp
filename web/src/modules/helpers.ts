import { OrderCommon } from './types';
import { MakerType } from './user/history';

const makerTypeMap = {
  ask: 'sell',
  bid: 'buy',
};

export const kindToMakerType = (kind: string): MakerType => makerTypeMap[kind];

export function getActualPrice(order: OrderCommon) {
  return order.price ?? order.avg_price ?? '0';
}
