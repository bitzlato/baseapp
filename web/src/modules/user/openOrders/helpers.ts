import { OrderCommon, OrderEvent } from '../../types';

export const convertOrderEvent = (orderEvent: OrderEvent): OrderCommon => {
  const { at, ...order } = orderEvent;

  return {
    ...order,
    ord_type: order.order_type || order.ord_type,
    created_at: typeof order.created_at === 'number' ? order.created_at * 1000 : order.created_at,
    updated_at: typeof order.updated_at === 'number' ? order.updated_at * 1000 : order.updated_at,
  };
};

export const insertOrUpdate = (list: OrderCommon[], order: OrderCommon): OrderCommon[] => {
  switch (order.state) {
    case 'pending':
    case 'wait':
      if (!list.find((d) => isSame(d, order))) {
        return [{ ...order }, ...list];
      }
      return list.map((d) => (isSame(d, order) ? { ...order } : d));

    default:
      return list.filter((d) => !isSame(d, order));
  }
};

function isSame(a: OrderCommon, b: OrderCommon): boolean {
  return (a.uuid !== undefined && b.uuid !== undefined && a.uuid === b.uuid) || a.id === b.id;
}
