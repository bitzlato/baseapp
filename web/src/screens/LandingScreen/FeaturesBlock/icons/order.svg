<svg id="Light_1" data-name="Light 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 102.25 100"><defs><style>.cls-2 { fill: #f84960; }
</style></defs><path d="M1398.88 1620a50 50 0 11-50-50 50 50 0 0150 50z" transform="translate(-1298.88 -1570)" fill="#eebe13"/><path class="cls-2" d="M1343.57 1659.59v10.13a49.79 49.79 0 01-29.72-14 4 4 0 011.15-.17h24.49a4.08 4.08 0 014.08 4.04zm-4.08-44.9H1315a4.09 4.09 0 00-4.08 4.09v24.49a4.08 4.08 0 004.08 4.08h24.49a4.08 4.08 0 004.08-4.08v-24.49a4.09 4.09 0 00-4.08-4.09zm0-40.81H1315a4.08 4.08 0 00-4.08 4.08v24.49a4.08 4.08 0 004.08 4.08h24.49a4.08 4.08 0 004.08-4.08V1578a4.08 4.08 0 00-4.08-4.12z" transform="translate(-1298.88 -1570)"/><path class="cls-2" d="M1343.57 1618.78v24.49a4.08 4.08 0 01-4.08 4.08h-16.33a4.08 4.08 0 01-4.08-4.08v-24.49a4.09 4.09 0 014.08-4.09h16.33a4.09 4.09 0 014.08 4.09zm-4.08-44.9h-16.33a4.08 4.08 0 00-4.08 4.08v24.49a4.08 4.08 0 004.08 4.08h16.33a4.08 4.08 0 004.08-4.08V1578a4.08 4.08 0 00-4.08-4.12zM1353.78 1657.55a2 2 0 11-2-2 2 2 0 012 2zm4.08-2a2 2 0 102 2 2 2 0 00-2-2.04zm6.12 0a2 2 0 102 2 2 2 0 00-1.98-2.04zM1401.12 1630v2a1 1 0 01-1 1h-47.34a1 1 0 01-1-1v-2a1 1 0 011-1h47.34a1 1 0 011 1zm-49.39-46.94v-2a1 1 0 011-1h36.73a1 1 0 011 1v2a1 1 0 01-1 1h-36.73a1 1 0 01-1-1zm1 9.19h40.81a1 1 0 001-1v-2a1 1 0 00-1-1h-40.81a1 1 0 00-1 1v2a1 1 0 001.03 1zm0 8.16h24.48a1 1 0 001-1v-2a1 1 0 00-1-1h-24.48a1 1 0 00-1 1v2a1 1 0 001.03 1zm0 24.49h36.73a1 1 0 001-1v-2a1 1 0 00-1-1h-36.73a1 1 0 00-1 1v2a1 1 0 001.03 1zm24.48 12.24h-24.48a1 1 0 00-1 1v2a1 1 0 001 1h24.48a1 1 0 001-1v-2a1 1 0 00-.97-1z" transform="translate(-1298.88 -1570)"/><path d="M1344.9 1574.91l2.28 2.29a1.61 1.61 0 010 2.28l-22.18 22.15a3.24 3.24 0 01-4.57 0l-11.11-11.12a1.61 1.61 0 010-2.28l2.29-2.29a1.61 1.61 0 012.28 0l7.69 7.69a1.61 1.61 0 002.28 0l18.72-18.72a1.62 1.62 0 012.32 0zm-2.29 40.82l-18.72 18.72a1.61 1.61 0 01-2.28 0l-7.69-7.69a1.61 1.61 0 00-2.28 0l-2.29 2.28a1.62 1.62 0 000 2.29l11.11 11.11a3.24 3.24 0 004.57 0l22.15-22.14a1.62 1.62 0 000-2.29l-2.28-2.28a1.62 1.62 0 00-2.29 0z" transform="translate(-1298.88 -1570)" fill="#f0f3f9"/></svg>
