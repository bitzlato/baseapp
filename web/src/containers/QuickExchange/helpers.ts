import type { Currency } from '@bitzlato/money-js';
import { lowerBound } from 'src/helpers/lowerBound';
import type { Market, Wallet } from '../../modules';

export interface DropdownItem {
  code: string;
  match: boolean;
}

interface SwapState {
  fromList: DropdownItem[];
  toList: DropdownItem[];
  market?: Market;
}

export function getCurrencies(markets: Market[], base: string, quote: string): SwapState {
  let market: Market | undefined;
  const fromList: DropdownItem[] = [];
  const toList: DropdownItem[] = [];

  for (let m of markets) {
    const sell = m.base_unit === base && m.quote_unit === quote;
    if (sell || (m.base_unit === quote && m.quote_unit === base)) {
      market = m;
    }

    if (quote !== m.base_unit) insertUnique(fromList, m.base_unit, m.quote_unit === quote);
    if (quote !== m.quote_unit) insertUnique(fromList, m.quote_unit, m.base_unit === quote);

    if (base !== m.base_unit) insertUnique(toList, m.base_unit, m.quote_unit === base);
    if (base !== m.quote_unit) insertUnique(toList, m.quote_unit, m.base_unit === base);
  }

  fromList.sort(compareItems);
  toList.sort(compareItems);

  return {
    fromList,
    toList,
    market,
  };
}

function insertUnique(items: DropdownItem[], code: string, match: boolean) {
  const value = { code, match };
  const index = lowerBound(items, value, lessName);
  const found = index !== items.length && !lessName(value, items[index]);
  if (!found) {
    items.splice(index, 0, value);
  } else {
    const d = items[index];
    d.match ||= match;
  }
}

function lessName(a: DropdownItem, b: DropdownItem) {
  return a.code < b.code;
}

function compareItems(a: DropdownItem, b: DropdownItem): number {
  if (a.match !== b.match) {
    return Number(b.match) - Number(a.match);
  }
  return a.code === b.code ? 0 : a.code < b.code ? -1 : 1;
}

export function getWallet(currency: string, wallets: Wallet[]): Wallet | undefined {
  const lowerCurrency = currency.toLowerCase();
  return wallets.find((w) => w.currency.code.toLowerCase() === lowerCurrency);
}

export function getCurrency(code: Currency['code'], minorUnit: Currency['minorUnit']) {
  return {
    code,
    minorUnit,
  };
}

export function getItem(code: string, match: boolean): DropdownItem {
  return { code, match };
}
