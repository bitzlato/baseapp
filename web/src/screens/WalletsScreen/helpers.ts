import { WalletItemData } from 'src/components/WalletItem/WalletItem';
import { createCcy, createMoney, PENCE_CCY } from 'src/helpers/money';
import { GeneralBalance } from 'src/modules/account/types';
import { Wallet } from 'src/modules/user/wallets/types';
import { getCurrencySymbol } from 'web/src/helpers/getCurrencySymbol';

export function getList(wallets: Wallet[], balances: GeneralBalance[]): WalletItemData[] {
  const res: WalletItemData[] = [];

  for (let i = 0; i < wallets.length; i++) {
    const wallet = wallets[i]!;
    const ccy = wallet.currency;
    const currencyId = getCurrencySymbol(ccy);
    const balance = balances.find((d) => currencyId === d.currency_id);
    const balanceP2P = balance?.p2p_balance ? createMoney(balance.p2p_balance, ccy) : undefined;
    const balanceMarket = wallet.balance;
    const balanceTotal = balanceMarket.add(balanceP2P ?? createMoney(0, ccy));
    const locked = wallet.locked.add(
      balance?.p2p_hold ? createMoney(balance.p2p_hold, ccy) : createMoney(0, ccy),
    );
    res.push({
      name: wallet.name,
      currency: ccy.code,
      balanceTotal,
      balanceP2P,
      balanceMarket,
      locked,
      approximate: createMoney(wallet.price, PENCE_CCY).multiply(balanceTotal.toString()),
      hasTransfer: balance !== undefined && balance.p2p_balance !== null,
      index: i,
    });
  }

  // add p2p only currencies
  for (let i = 0; i < balances.length; i++) {
    const balance = balances[i]!;
    if (balance.market_balance == null) {
      const currencyId = balance.currency_id;
      const wallet = wallets.find((d) => getCurrencySymbol(d.currency) === currencyId);
      if (!wallet && balance.p2p_balance && balance.p2p_hold) {
        res.push({
          name: currencyId,
          currency: currencyId,
          balanceTotal: createMoney(balance.p2p_balance, createCcy(currencyId, 8)),
          balanceP2P: createMoney(balance.p2p_balance, createCcy(currencyId, 8)),
          balanceMarket: undefined,
          locked: createMoney(balance.p2p_hold, createCcy(currencyId, 8)),
          approximate: createMoney(0, PENCE_CCY),
          hasTransfer: false,
          index: wallets.length + i,
        });
      }
    }
  }

  return res.sort((a, b) => {
    if (a.hasTransfer !== b.hasTransfer) {
      return Number(b.hasTransfer) - Number(a.hasTransfer);
    }
    return a.index - b.index;
  });
}
