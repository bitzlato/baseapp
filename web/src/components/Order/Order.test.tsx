import { Button } from 'react-bootstrap';
import { mount, shallow, ShallowWrapper } from 'enzyme';
import { TestComponentWrapper } from 'src/lib/test/wrapper';
import { DEFAULT_ORDER_TYPES } from 'src/helpers/order';
import { Order, OrderComponentProps } from '.';

// tslint:disable:no-magic-numbers

const defaultProps: OrderComponentProps = {
  onSubmit: jest.fn(),
  priceMarketBuy: 5,
  priceMarketSell: 10,
  currentMarketAskPrecision: 4,
  currentMarketBidPrecision: 5,
  availableBase: 200,
  availableQuote: 12,
  from: 'btc',
  to: 'eth',
  asks: [['10', '1']],
  bids: [['10', '1']],
  translate: jest.fn(),
  orderTypes: DEFAULT_ORDER_TYPES,
  minAmount: '0',
};

const setup = (props: Partial<OrderComponentProps> = {}) =>
  shallow(
    <TestComponentWrapper>
      <Order {...{ ...defaultProps, ...props }} />
    </TestComponentWrapper>,
  );

describe('Order', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('should match snapshot', () => {
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('button should be disabled', () => {
    // tslint:disable: no-shadowed-variable
    const wrapper = mount(
      <TestComponentWrapper>
        <Order {...{ ...defaultProps }} />
      </TestComponentWrapper>,
    );
    wrapper.find(Button).at(0).simulate('click');
    expect(defaultProps.onSubmit).toHaveBeenCalledTimes(0);
    wrapper.unmount();
  });
});
