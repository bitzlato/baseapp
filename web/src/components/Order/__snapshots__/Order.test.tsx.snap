// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order should match snapshot 1`] = `
<div
  class="cr-order cr-order--extended row alignStart rowSpacing"
>
  <div
    class="cr-order--extended__buy"
  >
    <div
      class="cr-tab-panel cr-tab-panel__fixed"
    >
      <div
        class="cr-tab-panel__navigation-container draggable-container"
      >
        <div
          class="cr-tab-panel__navigation-container-navigation"
          role="tablist"
        >
          <div
            class="cr-tab cr-tab__active"
            role="tab"
            tabindex="0"
          >
            Buy
            <span
              class="cr-tab__pointer"
            />
          </div>
        </div>
      </div>
      <div
        class="cr-tab-content cr-tab-content__active"
      >
        <div
          class="cr-order-form"
        >
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-item__dropdown__label"
            >
              Order Type
            </div>
            <div
              class="selfStretch positionRelative"
            >
              <label
                class="reactSelectLabel ellipsis"
              />
              <div
                class="reactSelect reactSelectSmall css-b62m3t-container"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="reactSelect__control css-1s2u09g-control"
                >
                  <div
                    class="reactSelect__value-container reactSelect__value-container--has-value css-319lph-ValueContainer"
                  >
                    <div
                      class="reactSelect__single-value css-qc6sy-singleValue"
                    >
                      Market
                    </div>
                    <div
                      class="reactSelect__input-container css-6j8wv5-Input"
                      data-value=""
                    >
                      <input
                        aria-autocomplete="list"
                        aria-controls="react-select-2-listbox"
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-owns="react-select-2-listbox"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class="reactSelect__input"
                        id="react-select-2-input"
                        role="combobox"
                        spellcheck="false"
                        style="label:input;color:inherit;background:0;opacity:1;width:100%;grid-area:1 / 2;font:inherit;min-width:2px;border:0;margin:0;outline:0;padding:0"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="reactSelect__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="reactSelect__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="reactSelect__indicator reactSelect__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="input row grow padding"
            >
              <label
                class="inputLabel"
              >
                Price
              </label>
              <div
                class="row grow rowSpacingSm"
              >
                <span>
                  ≈
                </span>
                <span
                  class="grow primaryColor ellipsis bold"
                >
                  5
                  <span
                    class="fractional"
                  >
                    .00
                  </span>
                </span>
                <span>
                  BTC
                </span>
              </div>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-item__prices"
            >
              <span
                class="smSize bidColor"
              >
                ▲ 10
                <span
                  class="fractional"
                >
                  .00
                </span>
              </span>
              <span
                class="smSize askColor"
              >
                ▼ 10
                <span
                  class="fractional"
                >
                  .00
                </span>
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-input"
            >
              <div
                class="custom-input"
              >
                <label />
                <input
                  class="form-control"
                  placeholder="Amount"
                  type="number"
                  value=""
                />
              </div>
              <span
                class="cr-order-input__currency"
              >
                ETH
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="row wrap justifyBetween selfStretch"
            >
              <label>
                Min Amount
              </label>
              <span
                class=""
              >
                <span
                  class="primaryColor"
                >
                  0
                  <span
                    class="fractional"
                  >
                    .00
                  </span>
                </span>
                 
                <span>
                  ETH
                </span>
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-item__percentage-buttons"
            >
              <button
                class="cr-percentage-button"
              >
                25%
              </button>
              <button
                class="cr-percentage-button"
              >
                50%
              </button>
              <button
                class="cr-percentage-button"
              >
                75%
              </button>
              <button
                class="cr-percentage-button"
              >
                100%
              </button>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="row wrap justifyBetween selfStretch lgSize"
            >
              <label>
                Total
              </label>
              <div
                class="row rowSpacingSm"
              >
                <span
                  class="secondaryColor"
                >
                  ≈
                </span>
                <span
                  class=""
                >
                  <span
                    class="primaryColor"
                  >
                    0
                    <span
                      class="fractional"
                    >
                      .00
                    </span>
                  </span>
                   
                  <span>
                    BTC
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="row wrap justifyBetween selfStretch"
            >
              <label>
                Available
              </label>
              <span
                class=""
              >
                <span
                  class="primaryColor"
                >
                  12
                  <span
                    class="fractional"
                  >
                    .00
                  </span>
                </span>
                 
                <span>
                  BTC
                </span>
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <button
              class="row rowSpacing btn btn-success btn-block btn-lg"
              disabled=""
              type="button"
            >
              <span>
                Buy
              </span>
              <span>
                ETH
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="cr-order--extended__sell"
  >
    <div
      class="cr-tab-panel cr-tab-panel__fixed"
    >
      <div
        class="cr-tab-panel__navigation-container draggable-container"
      >
        <div
          class="cr-tab-panel__navigation-container-navigation"
          role="tablist"
        >
          <div
            class="cr-tab cr-tab__active"
            role="tab"
            tabindex="0"
          >
            Sell
            <span
              class="cr-tab__pointer"
            />
          </div>
        </div>
      </div>
      <div
        class="cr-tab-content cr-tab-content__active"
      >
        <div
          class="cr-order-form"
        >
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-item__dropdown__label"
            >
              Order Type
            </div>
            <div
              class="selfStretch positionRelative"
            >
              <label
                class="reactSelectLabel ellipsis"
              />
              <div
                class="reactSelect reactSelectSmall css-b62m3t-container"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-3-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="reactSelect__control css-1s2u09g-control"
                >
                  <div
                    class="reactSelect__value-container reactSelect__value-container--has-value css-319lph-ValueContainer"
                  >
                    <div
                      class="reactSelect__single-value css-qc6sy-singleValue"
                    >
                      Market
                    </div>
                    <div
                      class="reactSelect__input-container css-6j8wv5-Input"
                      data-value=""
                    >
                      <input
                        aria-autocomplete="list"
                        aria-controls="react-select-3-listbox"
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-owns="react-select-3-listbox"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class="reactSelect__input"
                        id="react-select-3-input"
                        role="combobox"
                        spellcheck="false"
                        style="label:input;color:inherit;background:0;opacity:1;width:100%;grid-area:1 / 2;font:inherit;min-width:2px;border:0;margin:0;outline:0;padding:0"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="reactSelect__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="reactSelect__indicator-separator css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="reactSelect__indicator reactSelect__dropdown-indicator css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="input row grow padding"
            >
              <label
                class="inputLabel"
              >
                Price
              </label>
              <div
                class="row grow rowSpacingSm"
              >
                <span>
                  ≈
                </span>
                <span
                  class="grow primaryColor ellipsis bold"
                >
                  10
                  <span
                    class="fractional"
                  >
                    .00
                  </span>
                </span>
                <span>
                  BTC
                </span>
              </div>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-item__prices"
            >
              <span
                class="smSize bidColor"
              >
                ▲ 10
                <span
                  class="fractional"
                >
                  .00
                </span>
              </span>
              <span
                class="smSize askColor"
              >
                ▼ 10
                <span
                  class="fractional"
                >
                  .00
                </span>
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-input"
            >
              <div
                class="custom-input"
              >
                <label />
                <input
                  class="form-control"
                  placeholder="Amount"
                  type="number"
                  value=""
                />
              </div>
              <span
                class="cr-order-input__currency"
              >
                ETH
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="row wrap justifyBetween selfStretch"
            >
              <label>
                Min Amount
              </label>
              <span
                class=""
              >
                <span
                  class="primaryColor"
                >
                  0
                  <span
                    class="fractional"
                  >
                    .00
                  </span>
                </span>
                 
                <span>
                  ETH
                </span>
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="cr-order-item__percentage-buttons"
            >
              <button
                class="cr-percentage-button"
              >
                25%
              </button>
              <button
                class="cr-percentage-button"
              >
                50%
              </button>
              <button
                class="cr-percentage-button"
              >
                75%
              </button>
              <button
                class="cr-percentage-button"
              >
                100%
              </button>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="row wrap justifyBetween selfStretch lgSize"
            >
              <label>
                Total
              </label>
              <div
                class="row rowSpacingSm"
              >
                <span
                  class="secondaryColor"
                >
                  ≈
                </span>
                <span
                  class=""
                >
                  <span
                    class="primaryColor"
                  >
                    0
                    <span
                      class="fractional"
                    >
                      .00
                    </span>
                  </span>
                   
                  <span>
                    BTC
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <div
              class="row wrap justifyBetween selfStretch"
            >
              <label>
                Available
              </label>
              <span
                class=""
              >
                <span
                  class="primaryColor"
                >
                  200
                  <span
                    class="fractional"
                  >
                    .00
                  </span>
                </span>
                 
                <span>
                  ETH
                </span>
              </span>
            </div>
          </div>
          <div
            class="cr-order-item"
          >
            <button
              class="row rowSpacing btn btn-danger btn-block btn-lg"
              disabled=""
              type="button"
            >
              <span>
                Sell
              </span>
              <span>
                ETH
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
